File: src/main/java/com/ptah/service/user/UserService.java
package com.ptah.service.user;

import cn.hutool.core.lang.Validator;
import com.ptah.dto.userprofiling.RegisterRequest;
import com.ptah.dto.userprofiling.UserDto;
import com.ptah.entity.userprofiling.AuthorityMapping;
import com.ptah.entity.userprofiling.OrganizationNomination;
import com.ptah.entity.userprofiling.ProjectNomination;
import com.ptah.entity.userprofiling.ProjectRole;
import com.ptah.entity.userprofiling.User;
import com.ptah.repository.project.ProjectNominationRepository;
import com.ptah.repository.userprofiling.AuthorityMappingRepository;
import com.ptah.repository.userprofiling.OrganizationNominationRepository;
import com.ptah.repository.userprofiling.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
public class UserService {
    @Autowired
    private UserRepository userRepository;
    @Autowired
    private AuthorityMappingRepository authorityMappingRepository;
    @Autowired
    private ProjectNominationRepository projectNominationRepository;
    @Autowired
    private OrganizationNominationRepository organizationNominationRepository;

    public UserDto register(RegisterRequest registerRequest) {
        validateRegisterRequest(registerRequest);
        return createUser(registerRequest);
    }

    private UserDto createUser(RegisterRequest registerRequest) {
        User savedUser = userRepository.save(User.builder().login(registerRequest.getLogin()).password(registerRequest.getPassword()).build());
        return UserDto.builder().login(savedUser.getLogin()).build();
    }

    private void validateRegisterRequest(RegisterRequest registerRequest) {
        Validator.validateNotEmpty(registerRequest.getLogin(), "");
        Validator.validateNotEmpty(registerRequest.getPassword(), "");
    }

    public Set<GrantedAuthority> getAuthorities(User user) {
        Set roles = new HashSet();
        roles.addAll(projectNominationRepository.findByUserId(user.getId()).stream().map(ProjectNomination::getProjectRole).map(ProjectRole::name).toList());
        roles.addAll(organizationNominationRepository.findByUserId(user.getId()).stream().map(OrganizationNomination::getOrganizationRole).toList());
        authorityMappingRepository.findByRoles(roles);
        return convertToGrantedAuthorities(authorityMappingRepository.findByRoles(roles));
    }

    private Set<GrantedAuthority> convertToGrantedAuthorities(Set<AuthorityMapping> authorityMappings) {
        return authorityMappings.stream().map(authorityMapping -> new SimpleGrantedAuthority(authorityMapping.getRole())).collect(Collectors.toSet());
    }
}

File: backend/src/main/kotlin/com/ptah/service/userprofiling/UserService.kt
package com.ptah.service.userprofiling

import cn.hutool.core.lang.Validator
import com.ptah.common.*
import com.ptah.common.exceptions.ApplicationException
import com.ptah.dto.userprofiling.RegisterRequest
import com.ptah.dto.userprofiling.UserDto
import com.ptah.dto.userprofiling.UserInfo
import com.ptah.entity.userprofiling.*
import com.ptah.repository.project.ProjectNominationRepository
import com.ptah.repository.userprofiling.*
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.stereotype.Service
import kotlin.collections.HashSet

@Service
class UserService(
    private var userRepository: UserRepository,
    private var userJournalRepository: UserJournalRepository,
    private var organizationRepository: OrganizationRepository,
    private var authorityMappingRepository: AuthorityMappingRepository,
    private var projectNominationRepository: ProjectNominationRepository,
    private var organizationNominationService: OrganizationNominationService,
    private var organizationNominationRepository: OrganizationNominationRepository


) {
    fun register(registerRequest: RegisterRequest): UserDto {
        validateRegisterRequest(registerRequest)
        return createUser(registerRequest)
    }

    private fun createUser(registerRequest: RegisterRequest) = UserDto(
        login = userRepository.save(
            User(
                password = registerRequest.password, login = registerRequest.login
            )
        ).login
    )

    private fun validateRegisterRequest(registerRequest: RegisterRequest) {
        Validator.validateNotEmpty(registerRequest.login, "")
        Validator.validateNotEmpty(registerRequest.password, "")
    }

    fun getAuthorities(user: User): Set<GrantedAuthority> {
        val roles = getRolesOfUser(user)
        val authorityMappings = getUserAuthorityMappings(roles)
        val authorities = convertToGrantedAuthorities(authorityMappings)
        authorities.addAll(roles.map { SimpleGrantedAuthority(it) })
        return authorities
    }

    private fun getUserAuthorityMappings(roles: Set<String>): Set<AuthorityMapping?> {
        return authorityMappingRepository.findByRoles(roles)
    }

    private fun getRolesOfUser(user: User): MutableSet<String> {
        val roles: MutableSet<String> = with(HashSet<String>()) {
            addAll(projectNominationRepository.findByUserId(user.id).mapNotNull {
                it?.projectRole?.name?.lowercase()
            })
            addAll(organizationNominationRepository.findByUserId(user.id).mapNotNull {
                it?.organizationRole?.name?.lowercase()
            })
            this
        }
        return roles
    }

    private fun convertToGrantedAuthorities(authorityMappings: Set<AuthorityMapping?>): MutableSet<GrantedAuthority> {
        return authorityMappings.filterNotNull().flatMap { it.authorities!! }.map { SimpleGrantedAuthority(it) }
            .toMutableSet()
    }

    fun assignUserToOrganization(userId: Long, organizationId: Long, organizationRole: OrganizationRole) {
        val organizationNomination = organizationNominationService.getOrganizationNomination(userId, organizationId)
        organizationNominationService.updateOrganizationNomination(
            organizationNomination, userId, organizationId, organizationRole
        )
        organizationNominationRepository.save(organizationNomination)
    }


    fun switchToOrganization(userId: Long, organizationId: Long) {
        val user: User = getUser(userId)
        user.currentOrganization = organizationRepository.findById(organizationId)
            .orElseThrow { ApplicationException.of(Errors.ORGANIZATION_NOT_FOUND) }
        userRepository.save(user)
    }

    fun getUserInfo(userId: Long): UserInfo {
        val user: User = getUser(userId)
        val authorities = getAuthorities(user).map { it.authority }
        val userJournals = userJournalRepository.findByUserId(userId)
        return UserInfo().apply {
            this.user = UserDto().apply {
                this.authorities = authorities
                login = user.login
                nickname = user.name
            }
            logs = userJournals
        }
    }

    private fun getUser(userId: Long): User {
        return userRepository.findById(userId).orElseThrow { ApplicationException.of(Errors.USER_NOT_FOUND) }!!
    }

    fun updateUser(userDto: UserDto): UserDto {
        userRepository.save(userDto.toEntity(User::class))
        return userDto
    }

}